pkg/test/framework/components/apps/apps.go

```diff
@@ -57,8 +55,25 @@ type App interface {
 	Name() string
 	Endpoints() []AppEndpoint
 	EndpointsForProtocol(protocol model.Protocol) []AppEndpoint
-	Call(e AppEndpoint, opts AppCallOptions) ([]*echo.ParsedResponse, error)
-	CallOrFail(e AppEndpoint, opts AppCallOptions, t testing.TB) []*echo.ParsedResponse
+	/*
+		GetResponseAndError makes a call to `e` endpoint based on the protocol from `opts`.
+		It returns both the response and error regardless of they exist. The caller is expected to handle any errors.
+		This is needed mostly for Istio RBAC. When a TCP request is rejected due to authorization, the response
+		body will be empty. This doesn't mean that the request is not successful. If the authorization
+		policy is to reject this request, this is considered successful. In addition, in some cases, such as
+		when a request is expected to be rejected due to authorization, the response code is 403 (instead of 200),
+		and this is expected/considered successful.
+	*/
+	GetResponseAndError(e AppEndpoint, opts AppCallOptions) ([]*echo.ParsedResponse, *endpoint, error)
```

Comment:
I don't think this is an improvement. If you need `Call` to optionally return both, just do that.

Reply:
I think this is more flexible, no? We can alternatively add an extra parameter in the old `Call()` function to indicate whether the caller wants the framework to handle the error. However, not many callers are gonna use this extra parameter I'd imagine, some may want to have error handling by the test framework by default in most cases. So I don't think we should ask them to include an extra argument every time. Rather, we can have a separate function to do that.
 
Also, I think changing the names makes it more clear of what the functions are doing.

Reply:
I agree, it is a little convoluted. Get response and error, get response or error. Seems much simpler to just have a Call and a CheckCallSuccess or something. Then you can do `CheckCallSuccess(Call(...))`. Or have two methods, one that does a Call and one that does a Call and validates it.

Also minor point but this isn't RBAC specific, we test negatives for all sorts of use cases

Reply:
Just returning both in `Call` is better, it's kind hard to remember all these `GetResponseXXX`.

Reply:
fwiw @ozevren was against having Call return err and a response in https://github.com/istio/istio/pull/13099#discussion_r272735221.

Briefly looked around the web and didn't see anything related to whether response/err should be mutually exclusive or if it is ok to return both, so doesn't seem to be a established convention (that i could find)

Reply:
Many test frameworks in google3 have get<something>OrError, get<something>OrFail, etc. Personally, I like descriptive function names. `Call` doesn't tell me what it is, e.g. is it a callback function, etc. I'll leave the final decision to the test framework owners @nmittler @ozevren. Please advise on how to do this. So sum it up:
Currently:
* Call: makes some protocol call and return resp or error.
* CallOrFail: calls Call and fails or returns resp.

My proposal:
* GetResponseOrError (name change, replaces Call)
* GetResponseOrFail (name change, replaces CallorFail)
* GetResponseAndError: makes some protocol call and return resp and error. Callers must handle error message if they decide to use this. Needed for RBAC and negatives test (as John added).

Other proposals...

Reply:
I am having trouble understanding the subtlety here. Return of error indicates something is borked pretty badly. If it is a multi-value return with an error, it means the call has failed (horribly) and the returned value is not dependable. This is just general call convention most people expect when writing code. If you're returning a result, then it means things have succeeded.

If you need to model some "partial success", then the right way to model is by changing the return value. For example, if you were to write a compiler, where there is partial success, or detailed error information, you wouldn't implement this as a result & error, but as a single result object that contains both the result of partial parsing, as well a set of detailed diagnostic information.

Similarly, if you have a case of multiple different types of use cases, you'd have different overloads, or additional methods for doing further validation. Which seems to be the most applicable use case here.

So, the model would be similar to this:
```golang
// Call and return the result (which may have protocol level errors, such as HTTP codes even in the case of success), or error if the call was not made (i.e. connection refused, underwater cables are cut etc.).
func Call() (CallResult, error)

// Call and expect success, otherwise simply fail the test. This is just a convenience to avoid writing a whole if block when calling Call()
func CallOrFail() (CallResult, error)

// The additional validation can be handled as additional validation check calls
func (c CallResult) IsValidForSomeConstraint() bool
func (c CallResult) isErrorCode(errCode) bool 
func (c CallResult) EnsureErrorCode(t *testing.T, errorCode)
// etc.
```

Reply:
Ok so looks like we are moving forward with John's PR here (https://github.com/istio/istio/pull/13099). Thank you all